lib/Corres_UL.thy

```diff
@@ -537,6 +537,13 @@ lemmas corres_when2 =
                 and Q="\<top>" and Q'="\<top>" and r=dc, unfolded if_apply_def2, simplified,
                 folded when_def]
 
+lemma corres_if4:
+  "\<lbrakk> G = G';
+     G \<Longrightarrow> corres_underlying sr nf nf' r P P' a c;
+     \<not>G \<Longrightarrow> corres_underlying sr nf nf' r P P' b d \<rbrakk>
+   \<Longrightarrow> corres_underlying sr nf nf' r P P' (if G then a else b) (if G' then c else d)"
+  by (clarsimp simp: corres_underlying_def)
+
```

Comment:
Hmm. This looks quite trivial to me. `simp` or `clarsimp` will split the conclusion into a conjunction of the two implications in the assumption or something similar. So, `clarsimp` and `conjI/impI` would be pretty much the same thing as applying this rule, woundn’t it?

Reply:
It is quite trivial. `clarsimp` seems to be greedy and will pull apart all `if`s. `schedule_used` contains three `if`s, so here it will create duplication which I can't see how to handle easily, and disturbs the structure of the proof I was wanting

Reply:
Ok, you want to proceed one layer at a time. Did you try using `corres_guard_imp` and `corres_if3`? Or `corres_guard_cong`?

Reply:
I could use `corres_guard_imp` and `corres_if3`, but I was aiming for a more "fowards" style, where the guards are either provided by strongest postcondition rules, "skip" rules, comparison with the initial guards, or (as a last resort) manual instantiation. `corres_if4` is exactly 

```
corres_if3[where P=T and Q=T for T, where P'=T' and Q'=T' for T', simplified]
```

so I could just use that instead, but I thought that a simple rule like this might be more convenient for people who like this (even more) forwards reasoning style in `corres` proofs

Reply:
I would certainly prefer to keep it. `corres_if_split` isn't taken. Would you be happy for me to rename it to `corres_if_split`?

Reply:
that would be good, if there’s no objection from other people.